File: src/python/_0649_Dota2_Senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        rAlive = 0
        dAlive = 0
        guys = deque()

        for guy in senate:
            guys.append(guy)
            if guy == "R":
                rAlive += 1
            else:
                dAlive += 1

        rToKill = 0
        dToKill = 0
        while rAlive > 0 and dAlive > 0:
            currentGuy = guys.popleft()
            if currentGuy == "R":
                if rToKill > 0:
                    rToKill -= 1
                    rAlive -= 1
                else:
                    dToKill += 1
                    guys.append(currentGuy)
            else:
                if dToKill > 0:
                    dToKill -= 1
                    dAlive -= 1
                else:
                    rToKill += 1
                    guys.append(currentGuy)

        return "Radiant" if rAlive > 0 else "Dire"

File: src/_0649_Dota2_Senate.java
class Solution {
    public String predictPartyVictory(String senate) {
        int rAlive = 0, dAlive = 0;
        Queue<Character> guys = new LinkedList<>();

        for (char guy : senate.toCharArray()) {
            guys.add(guy);
            if (guy == 'R')
                rAlive++;
            else
                dAlive++;
        }

        int rToKill = 0, dToKill = 0;
        while (rAlive > 0 && dAlive > 0) {
            char currentGuy = guys.remove();
            if (currentGuy == 'R') {
                if (rToKill > 0) {
                    rToKill--;
                    rAlive--;
                } else {
                    dToKill++;
                    guys.add(currentGuy);
                }
            } else {
                if (dToKill > 0) {
                    dToKill--;
                    dAlive--;
                } else {
                    rToKill++;
                    guys.add(currentGuy);
                }
            }
        }

        if (rAlive > 0)
            return "Radiant";
        else
            return "Dire";
    }
}
